src/bionics.cpp

```diff
@@ -1705,6 +1707,92 @@ void bionics_install_failure( player *u, int difficulty, int success )
     }
 }
 
+std::string list_occupied_bps( std::string bio_id, std::string intro, bool one_per_line )
+{
+    std::ostringstream desc;
+    desc << intro;
+    for( const auto &elem : bionic_info( bio_id ).occupied_bodyparts ) {
+        desc << ( one_per_line ? "\n" : " " );
+        //~ <Bodypart name> (<number of occupied slots> slots);
+        desc << string_format( _( "%s (%i slots);" ),
+                           body_part_name_as_heading( elem.first, 1 ).c_str(),
+                           elem.second );
+    }
+    return desc.str();
+}
+
+int player::get_used_bionics_slots( const body_part bp ) const
+{
+    int used_slots = 0;
+    for( auto &bio : my_bionics ) {
+        auto search = bionics[bio.id].occupied_bodyparts.find( bp );
+        if( search != bionics[bio.id].occupied_bodyparts.end() ) {
+            used_slots += search->second;
+        }
+     }
+
+    return used_slots;
+}
+
+std::map<body_part, int> player::bionic_installation_issues( std::string bioid )
+{
+    std::map<body_part, int> issues;
+    for( auto& elem : bionics[ bioid ].occupied_bodyparts ) {
+        if( !has_enough_slots( elem.first, elem.second ) ) {
+            issues.emplace( elem.first, (int)elem.second -
+                            get_free_bionics_slots( elem.first ) );
+        }
```

Comment:
I'd rewrite the inner snippet like this:
```c++
const int lacked_slots = elem.second - get_free_bionics_slots( elem.first );
if( lacked_slots > 0 ) {
    issues.emplace( elem.first, lacked_slots );
}
```
It's slightly more expressive, saves you one redundant call to `get_free_bionics_slots()`, and enables getting rid of `has_enough_slots()`.